### PR TITLE
Parser refactor

### PR DESCRIPTION
**Pull request by chfast**

### CODE REVIEW

File: lib/fizzy/parser.cpp [hunk 1]
```diff
@@ -438,4 +464,9 @@ Module parse(bytes_view input)
 
     return module;
 }
+
+parser_result<std::vector<uint32_t>> parse_i32s(const uint8_t* pos, const uint8_t* end)
```
Comment (axic) [inline lib/fizzy/parser.cpp]: Why not `parse_i32_vec` or `parse_vec_i32`?

File: lib/fizzy/parser.hpp [hunk 2]
```diff
@@ -51,23 +45,5 @@ inline parser_result<Limits> parse_limits(const uint8_t* pos, const uint8_t* end
...
-    //       appearing in big set of example wasm binaries.
-    if (size < 128)
-        result.reserve(size);
-
-    auto inserter = std::back_inserter(result);
-    for (uint32_t i = 0; i < size; ++i)
-        std::tie(inserter, pos) = parse<T>(pos, end);
-    return {result, pos};
-}
+parser_result<std::vector<uint32_t>> parse_i32s(const uint8_t* pos, const uint8_t* end);
```
Comment (axic) [inline lib/fizzy/parser.hpp]: Perhaps a comment noting this is only used by `parse_expr.cpp` would be useful?

File: lib/fizzy/parser.cpp [hunk 3]
```diff
@@ -64,6 +64,12 @@ parser_result<ValType> parse(const uint8_t* pos, const uint8_t* end)
     }
 }
 
+const uint8_t* validate_valtype(const uint8_t* pos, const uint8_t* end)
+{
+    const auto [_, next_pos] = parse<ValType>(pos, end);
```
Comment (axic) [inline lib/fizzy/parser.cpp]: Why is this better than `std::tie(std::ignore, pos) = parse...` ?

File: lib/fizzy/parser.cpp [hunk 4]
```diff
@@ -64,6 +64,12 @@ parser_result<ValType> parse(const uint8_t* pos, const uint8_t* end)
     }
 }
 
+const uint8_t* validate_valtype(const uint8_t* pos, const uint8_t* end)
```
Comment (axic) [inline lib/fizzy/parser.cpp]: Shouldn't this be inline?

**Comment by axic** [comment]
There seems to still be a few unnecessarily exported parser helpers. Anything to move in this PR?

File: lib/fizzy/parser.cpp (hunk 4 above)
Comment (chfast) [reply lib/fizzy/parser.cpp]: It cannot, because is used in other compilation unit.

File: lib/fizzy/parser.cpp (hunk 3 above)
Comment (chfast) [reply lib/fizzy/parser.cpp]: Can be, but it will weirdly update the input argument.

File: lib/fizzy/parser.cpp (hunk 4 above)
Comment (axic) [reply lib/fizzy/parser.cpp]: Oh thought this was the header. Nvm.

**Comment by chfast** [comment]
> There seems to still be a few unnecessarily exported parser helpers. Anything to move in this PR?

Do you mean `parse_limits` and `parse_string`?

First is used in unittests. I can replace it with full wasm tests, but @gumb0 had some restrictions.

Second is used in microbenchmarking. Implementation is not exposed.

**Comment by axic** [comment]
> Do you mean parse_limits and parse_string?

To be honest haven't checked the file before commenting. In the past we had *a lot* in there. Checked now, you are right, it is mostly `parse_limits` which would annoy me.

**Pull request merged**